File: scribbleshare-room/src/main/java/net/stzups/scribbleshare/room/server/websocket/protocol/client/messages/ClientMessageCanvasInsert.java
package net.stzups.scribbleshare.room.server.websocket.protocol.client.messages;

import io.netty.buffer.ByteBuf;
import net.stzups.scribbleshare.data.objects.canvas.CanvasInsert;
import net.stzups.scribbleshare.data.objects.canvas.object.CanvasObjectType;
import net.stzups.scribbleshare.room.server.websocket.protocol.client.ClientMessage;
import net.stzups.scribbleshare.room.server.websocket.protocol.client.ClientMessageType;

import java.util.HashMap;
import java.util.Map;

public class ClientMessageCanvasInsert extends ClientMessage {
    private final Map<CanvasObjectType, CanvasInsert[]> canvasInsertsMap = new HashMap<>();

    public ClientMessageCanvasInsert(ByteBuf byteBuf) {
        super(ClientMessageType.CANVAS_INSERT);
        int length = byteBuf.readUnsignedByte();
        for (int i = 0; i < length; i++) {
            CanvasObjectType type = CanvasObjectType.valueOf(byteBuf.readUnsignedByte());
            CanvasInsert[] canvasInserts = new CanvasInsert[byteBuf.readUnsignedByte()];
            for (int j = 0; j < canvasInserts.length; j++) {
                canvasInserts[j] = new CanvasInsert(type, byteBuf);
            }
            canvasInsertsMap.put(type, canvasInserts);
        }
    }

    public Map<CanvasObjectType, CanvasInsert[]> getCanvasInsertsMap() {
        return canvasInsertsMap;
    }
}

File: scribbleshare-frontend/src/scripts/protocol/client/messages/ClientMessageCanvasInsert.js
import ClientMessageType from "../ClientMessageType.js";
import ClientMessage from "../ClientMessage.js";

export default class ClientMessageCanvasInsert extends ClientMessage {
    constructor(canvasInsertsMap) {
        super(ClientMessageType.CANVAS_INSERT);
        this.canvasInsertsMap = canvasInsertsMap;
    }

    serialize(writer) {
        super.serialize(writer);
        writer.writeUint8(this.canvasInsertsMap.size);
        this.canvasInsertsMap.forEach((canvasInserts, canvasObjectType) => {
            writer.writeUint8(canvasObjectType);
            writer.writeUint8(canvasInserts.length);
            canvasInserts.forEach((canvasInsert) => {
                canvasInsert.serialize(writer);
            });
        });
    }
}